AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Core/DataStoreListProvider.swift

```diff
@@ -0,0 +1,97 @@
+        /// hold the information necessary to query the associated elements (e.g. comments of a post)
+        ///
+        /// The associatedField represents the field to which the owner of the `List` is linked to.
+        /// For example, if `Post.comments` is associated with `Comment.post` the `List<Comment>`
+        /// of `Post` will have a reference to the `post` field in `Comment`.
+        case notLoaded(associatedId: Model.Identifier, associatedField: String)
+
+        case loaded([Element])
+    }
+
+    var loadedState: LoadedState
+
+    init(associatedId: Model.Identifier,
+         associatedField: String) {
+        self.loadedState = .notLoaded(associatedId: associatedId,
+                                      associatedField: associatedField)
+    }
+
+    init(_ elements: [Element]) {
+        self.loadedState = .loaded(elements)
+    }
+
+    public func load() -> Result<[Element], CoreError> {
+        let semaphore = DispatchSemaphore(value: 0)
+        var loadResult: Result<[Element], CoreError> = .failure(
+            .listOperation("DataStore query failed to complete.",
+                           AmplifyErrorMessages.shouldNotHappenReportBugToAWS(),
+                           nil))
+        load { result in
+            defer {
+                semaphore.signal()
+            }
+            switch result {
+            case .success(let elements):
+                loadResult = .success(elements)
+            case .failure(let error):
+                Amplify.DataStore.log.error(error: error)
+                assert(false, error.errorDescription)
+                loadResult = .failure(error)
+            }
+        }
+        semaphore.wait()
+        return loadResult
+    }
+
+    public func load(completion: (Result<[Element], CoreError>) -> Void) {
+        switch loadedState {
+        case .loaded(let elements):
+            completion(.success(elements))
+        case .notLoaded(let associatedId, let associatedField):
+            let predicate: QueryPredicate = field(associatedField) == associatedId
+            Amplify.DataStore.query(Element.self, where: predicate) {
+                switch $0 {
+                case .success(let elements):
+                    self.loadedState = .loaded(elements)
+                    completion(.success(elements))
+                case .failure(let error):
+                    Amplify.DataStore.log.error(error: error)
+                    completion(.failure(CoreError.listOperation("Failed to Query DataStore.",
+                                                                "See underlying DataStoreError for more details.",
+                                                                error)))
+                }
+            }
+        }
+    }
+
+    public func hasNextPage() -> Bool {
+        fatalError("Not implemented.")
+    }
+
+    public func getNextPage(completion: (Result<List<Element>, CoreError>) -> Void) {
+        fatalError("Not implemented.")
```

Comment:
Why fatalError instead of a `.failure` result like you did for the ArrayLiteral?

Reply:
In this PR, i wanted to keep the logic to a minimal, even the pagination APIs are not added to `List` type. I think i have added the actual implementation and testing in the next PR